certbot-ci/windows_installer_integration_tests/test_main.py

```diff
@@ -3,57 +3,211 @@
+                    self.send_response(200)
+                    self.send_header('Content-type', 'application/octet-stream')
+                    self.end_headers()
+                    with open(installer, 'rb') as file_h:
+                        self.wfile.write(file_h.read())
+                else:
+                    self.send_response(404)
+                    self.end_headers()
 
-        log_path = os.path.join('C:\\', 'Certbot', 'log', 'letsencrypt.log')
+        server_address = ('', 8009)
+        server = ThreadedTCPServer(server_address, GitHubMock)
+        thread = threading.Thread(target=server.serve_forever)
+        thread.daemon = True
+        thread.start()
+
+        yield 'http://localhost:8009/releases/latest'
+    finally:
+        if server:
+            server.shutdown()
+            server.server_close()
 
-        modification_time = os.path.getmtime(log_path)
-        assert now < modification_time, 'Certbot log file has not been modified by the renew task.'
 
-        with open(log_path) as file_h:
-            data = file_h.read()
-        assert 'no renewal failures' in data, 'Renew task did not execute properly.'
+@pytest.fixture
+def upgrade_env(signing_cert, github_mock):
+    """
+    This fixture prepares the current Windows system Registry for a proper blackbox testing
+    of the auto-upgrade mechanism. GitHub release API is set to use the local GitHub release
+    API mock. And the public key used to validate the installer is set to be the one from
+    the test signing certificate used in the tests.
+    Fixture cleanup is included.
+    """
+    try:
+        _ps('New-Item -Path HKLM:\\Software -Name Certbot -ErrorAction SilentlyContinue | Out-Null; exit 0')
+        _ps('New-ItemProperty -Path HKLM:\\Software\\Certbot -Name CertbotUpgradeApiURL -Value {} '
+            '| Out-Null'.format(github_mock))
+        _ps('New-ItemProperty -Path HKLM:\\Software\\Certbot -Name CertbotSigningPubKey -Value '
+            '([Convert]::ToBase64String((Get-PfxCertificate -FilePath {0}).GetPublicKey())) '
+            '| Out-Null'.format(signing_cert))
 
+        yield True
     finally:
-        # Sadly this command cannot work in non interactive mode: uninstaller will ask explicitly permission in an UAC prompt
-        # print('Uninstalling Certbot ...')
-        # uninstall_path = _ps('(gci "HKLM:\\SOFTWARE\\Wow6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall"'
-        #                      ' | foreach { gp $_.PSPath }'
-        #                      ' | ? { $_ -match "Certbot" }'
-        #                      ' | select UninstallString)'
-        #                      '.UninstallString', capture_stdout=True)
-        # subprocess.check_call([uninstall_path, '/S'])
+        _ps('Remove-ItemProperty -Path HKLM:\\Software\\Certbot -Name CertbotUpgradeApiURL')
+        _ps('Remove-ItemProperty -Path HKLM:\\Software\\Certbot -Name CertbotSigningPubKey')
```

Comment:
Come on, having useless entries remain when the uninstaller kicks in, then having to use CCleaner to clean you registry after 6 months of that process, is a national sport for Windows users!

More seriously, deleting the entire `HKLM:\\Software\Certbot` is not that obvious. For instance a lot of programs use the registry to store user data, and I expect Certbot to do it soon also. In particular, we will offer the possibility to choose what is the default config folder to use. This value should not be deleted without explicit user approval, and this will be necessary during the auto upgrade process, since we uninstall Certbot first, so we will need to have a way to remember what was the user config folder in previous installation.

For these reasons, I think it is better to delete explicitly only the keys that relates to the installation of Certbot itself, and not the user, and I think it is good to start using this approach now.